mantidproject/mantid
# Support d-spacing workspaces in IntegratePeaksSkew and improve peak validation

**Description of work.**

Several improvements to the `IntegratePeaksSkew` algorithm

1. Allow integration of workspaces in d-spacing (previously just TOF) - this required re-writing a lot of code to use mantid algorithms as opposed to numpy operations (due to non-aligned bin edges)
2. Add option not to optimise 1D window size below a threshold Intensity/sigma ratio
3. Improve formatting of output plot by putting legend outside the axes
4. Validate peaks based on number of statistically significant non-zero TOF bins

In addition a bug in estimating resolution parameters has been fixed

Note that as part of this PR this PR the error on empty bins is no longer increased to 1 (as is done in SXD2001) as this has been show to be invalid in issue #34841


**To test:**

(1) Run this script

```
Load(Filename='SXD26062.raw', OutputWorkspace='SXD26062')

CreatePeaksWorkspace(InstrumentWorkspace='SXD26062', NUmberOfPeaks=0, OutputWorkspace='SingleCrystalPeakTable')

AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=2366.8220873404448, DetectorID=34389, Height=1772, BinCount=95.229762433928016)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=2126.2022543400831, DetectorID=33804, Height=1505, BinCount=27.480506255931004)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=2184.0008770008217, DetectorID=35949, Height=1808, BinCount=52.661280309925587)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=1995.2115486292446, DetectorID=31765, Height=1513, BinCount=23.65730722975853)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=2252.1712809128194, DetectorID=30058, Height=2302, BinCount=84.449848622374006)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=5843.6050001340082, DetectorID=15121, Height=1639, BinCount=63.362132138439677)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=4419.0059861014552, DetectorID=38933, Height=885, BinCount=17.337073210576683)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=3023.5648640387335, DetectorID=37810, Height=710, BinCount=12.82320474223436)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=5145.6860467522974, DetectorID=38691, Height=1325, BinCount=126.70521154333336)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=4975.3450154521515, DetectorID=8558, Height=1345, BinCount=7.0523493183511672)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=16259.3, DetectorID=19474, Height=39, BinCount=6.5623281756867184)
 
ConvertUnits(InputWorkspace='SXD26062', Target='dSpacing',  OutputWorkspace='SXD26062')

fpath = '/mnt/babylon/Public/RWaite/SXD26062_peaks_int_skew.pdf'
IntegratePeaksSkew(InputWorkspace='SXD26062', PeaksWorkspace='SingleCrystalPeakTable', 
                   # NCols=3, NRows=3, NRowMax=3,NColMax=3,
                   BackscatteringTOFResolution=0.06, ThetaWidth=1.5, 
                   ScaleThetaWidthByWavelength=True, OptimiseMask=False, NVacanciesMax=1, 
                   UseNearestPeak=True, UpdatePeakPosition=False, 
                   OutputFile=fpath, 
                   LorentzCorrection=False, OutputWorkspace='SXD26062_peaks_int', 
                   OptimiseXWindowSize=True,
                   ThresholdIOverSigma=0.0)
```
(2) Inspect the output table - the peaks should be integrated
(3) Inspect the output pdf - the peak window should look like it covers the majority of the peaks (in 2D and 1D)
Fixes #35271







---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
